### PR TITLE
Issue #12748 fix

### PR DESCRIPTION
**Pull request by MaksimKiselev**
| Q | A |
| --- | --- |
| Is bugfix? | no |
| New feature? | yes |
| Breaks BC? | no |
| Tests pass? | yes |
| Fixed issues | #12748 |

### CODE REVIEW

File: framework/console/controllers/MigrateController.php [hunk 1]
```diff
@@ -166,9 +166,7 @@ public function optionAliases()
     public function beforeAction($action)
     {
         if (parent::beforeAction($action)) {
-            if ($action->id !== 'create') {
-                $this->db = Instance::ensure($this->db, Connection::className());
-            }
+            $this->db = Instance::ensure($this->db, Connection::className());
```
Comment (samdark) [inline framework/console/controllers/MigrateController.php]: I'd leave it as is here and duplicate the call in `create` action so it won't trigger errors if it doesn't need database.

File: framework/console/controllers/MigrateController.php [hunk 2]
```diff
@@ -312,11 +312,28 @@ protected function generateMigrationSourceCode($params)
             $table = mb_strtolower($matches[1], Yii::$app->charset);
         }
 
-        foreach ($foreignKeys as $column => $relatedTable) {
+        foreach ($foreignKeys as $column => $foreignKey) {
+            $relatedColumn = $foreignKey['column'];
+            $relatedTable = $foreignKey['table'];
+            if (empty($relatedColumn)) {
+                $relatedColumn = 'id';
+                $relatedTableSchema = $this->db->getTableSchema($relatedTable);
```
Comment (samdark) [inline framework/console/controllers/MigrateController.php]: You need DB here so call `$this->db = Instance::ensure($this->db, Connection::className());`.

File: framework/console/controllers/MigrateController.php [hunk 3]
```diff
@@ -312,11 +312,28 @@ protected function generateMigrationSourceCode($params)
...
+                if ($relatedTableSchema !== null) {
+                    $primaryKeyCount = count($relatedTableSchema->primaryKey);
+                    if ($primaryKeyCount > 1) {
+                        $this->stdout("Related table for field {$column} already exist, but primary key is a composite. Default name will be used for related field\n", Console::FG_YELLOW);
```
Comment (samdark) [inline framework/console/controllers/MigrateController.php]:
```php
"Related table for field \"{$column}\" exists, but primary key is composite. Default name \"id\" will be generated for related field\n"
```

File: framework/console/controllers/MigrateController.php [hunk 4]
```diff
@@ -312,11 +312,28 @@ protected function generateMigrationSourceCode($params)
...
+                    } elseif ($primaryKeyCount < 1) {
+                        $this->stdout("Related table for field {$column} already exist, but do not have primary key. Default name will be used for related field.\n", Console::FG_YELLOW);
```
Comment (samdark) [inline framework/console/controllers/MigrateController.php]:
```php
"Related table for field \"{$column}\" exists, but does not have a primary key. Default name \"id\" will be used for related field.\n"
```

File: framework/console/controllers/MigrateController.php [hunk 5]
```diff
@@ -363,10 +380,15 @@ protected function parseFields()
 
             foreach ($chunks as $i => &$chunk) {
                 if (strpos($chunk, 'foreignKey') === 0) {
-                    preg_match('/foreignKey\((\w*)\)/', $chunk, $matches);
-                    $foreignKeys[$property] = isset($matches[1])
-                        ? $matches[1]
-                        : preg_replace('/_id$/', '', $property);
+                    preg_match('/foreignKey\((\w*)\s?(\w*)\)/', $chunk, $matches);
```
Comment (samdark) [inline framework/console/controllers/MigrateController.php]: Why the regexp change?

File: framework/console/controllers/MigrateController.php (hunk 2 above)
Comment (MKiselev) [inline framework/console/controllers/MigrateController.php]: How to check that the component DB may be initialized here?

Я имею в виду, что если в конфигурации приложения db не задан, то DI не может разрешить эту зависимость и получаю исключение.

Comment (samdark) [inline framework/console/controllers/MigrateController.php]: https://github.com/yiisoft/yii2/pull/12770#discussion_r83872685

You can probably try-catch exception and fall back to default generation scenario...

File: framework/console/controllers/MigrateController.php (hunk 5 above)
Comment (MKiselev) [inline framework/console/controllers/MigrateController.php]: Regular expression changed for support setting field name for reference key by using whitespace like this: ```foreignKey(table_name reference_key)```

Now FK may by defined in three ways:
```table_id:foreignKey(table_name reference_key)```
```table_id:foreignKey(table_name)```
```table_id:foreignKey```

**Comment by MKiselev** [comment]
Видимо не так CHANGELOG.md надо править, я стянул с мастера и поправил...
Ещё и имя с битбакета в гите настроено.

**Comment by MKiselev** [comment]
@samdark have any comments for last commit?

File: framework/console/controllers/MigrateController.php [hunk 6]
```diff
@@ -293,8 +295,8 @@ protected function generateMigrationSourceCode($params)
 
             $foreignKeys[$firstTable . '_id']['table'] = $firstTable;
             $foreignKeys[$secondTable . '_id']['table'] = $secondTable;
-            $foreignKeys[$firstTable . '_id']['column'] = 'id';
-            $foreignKeys[$secondTable . '_id']['column'] = 'id';
+            $foreignKeys[$firstTable . '_id']['column'] = '';
```
Comment (samdark) [inline framework/console/controllers/MigrateController.php]: I'd still fall back to just `id`.

File: framework/console/controllers/MigrateController.php [hunk 7]
```diff
@@ -317,16 +319,21 @@ protected function generateMigrationSourceCode($params)
...
+                try {
+                    $this->db = Instance::ensure($this->db, Connection::className());
+                    $relatedTableSchema = $this->db->getTableSchema($relatedTable);
+                    if ($relatedTableSchema !== null) {
+                        $primaryKeyCount = count($relatedTableSchema->primaryKey);
+                        if ($primaryKeyCount > 1) {
+                            $this->stdout("Related table for field \"{$column}\" exists, but primary key is composite. Default name \"id\" will be generated for related field\n", Console::FG_YELLOW);
+                        } elseif ($primaryKeyCount < 1) {
+                            $this->stdout("Related table for field \"{$column}\" exists, but does not have a primary key. Default name \"id\" will be used for related field.\n", Console::FG_YELLOW);
+                        } else {
```
Comment (samdark) [inline framework/console/controllers/MigrateController.php]: This is the most common variant so I'd check for `=== 1` first.

File: framework/console/controllers/MigrateController.php [hunk 8]
```diff
@@ -317,16 +319,21 @@ protected function generateMigrationSourceCode($params)
...
-                    } else {
-                        $relatedColumn = $relatedTableSchema->primaryKey[0];
+                try {
+                    $this->db = Instance::ensure($this->db, Connection::className());
+                    $relatedTableSchema = $this->db->getTableSchema($relatedTable);
+                    if ($relatedTableSchema !== null) {
+                        $primaryKeyCount = count($relatedTableSchema->primaryKey);
+                        if ($primaryKeyCount > 1) {
+                            $this->stdout("Related table for field \"{$column}\" exists, but primary key is composite. Default name \"id\" will be generated for related field\n", Console::FG_YELLOW);
+                        } elseif ($primaryKeyCount < 1) {
```
Comment (samdark) [inline framework/console/controllers/MigrateController.php]: It's better to write it as `=== 0` since it can't be negative anyway.

**Review by samdark — CHANGES_REQUESTED** [review]

File: framework/console/controllers/MigrateController.php (hunk 6 above)
Comment (MKiselev) [inline framework/console/controllers/MigrateController.php]: @samdark  if here use just ```id```, then here https://github.com/MKiselev/yii2/blob/4126e46c3889373a54d59a22fe49c5b485d370c1/framework/console/controllers/MigrateController.php#L320 check will be false.

**Comment by MKiselev** [comment]
@samdark update code.

File: framework/console/controllers/MigrateController.php (hunk 6 above)
Comment (samdark) [inline framework/console/controllers/MigrateController.php]: Ah. OK then.

File: framework/console/controllers/MigrateController.php [hunk 9]
```diff
@@ -293,8 +293,10 @@ protected function generateMigrationSourceCode($params)
                 ]
             );
 
-            $foreignKeys[$firstTable . '_id'] = $firstTable;
-            $foreignKeys[$secondTable . '_id'] = $secondTable;
+            $foreignKeys[$firstTable . '_id']['table'] = $firstTable;
+            $foreignKeys[$secondTable . '_id']['table'] = $secondTable;
+            $foreignKeys[$firstTable . '_id']['column'] = '';
```
Comment (samdark) [inline framework/console/controllers/MigrateController.php]: Maybe it should be set to `null` and then the check could be ` === null`?

File: framework/console/controllers/MigrateController.php (hunk 5 above)
Comment (samdark) [inline framework/console/controllers/MigrateController.php]: Docs should be updated accordingly.

**Review by samdark — CHANGES_REQUESTED** [review]
Looks much better now. Added a bit more inline comments. Also need a changelog.

**Comment by MKiselev** [comment]
@samdark code was updated. My english is not very good, please check docs. RU docs I will update after accept PR.

File: framework/console/controllers/MigrateController.php (hunk 5 above)
Comment (MKiselev) [inline framework/console/controllers/MigrateController.php]: Was updated.

**Review by samdark — APPROVED** [review]
Everything's OK except wording.

**Comment by MKiselev** [comment]
@samdark What exactly is wrong?
I added RU docs.

**Comment by samdark** [comment]
Some wording in English. I can handle that...

**Comment by samdark** [comment]
Merged with some adjustments: https://github.com/yiisoft/yii2/commit/4f89265fff54723b6a2fa624a2f89429cd482d54. Thanks!

**Pull request closed without merging**